crates/merkle-tree/src/class.rs

```diff
@@ -33,8 +33,11 @@ impl<'tx> ClassCommitmentTree<'tx> {
     }
 
     /// Applies and persists any changes. Returns the new global root.
-    pub fn apply(self) -> anyhow::Result<ClassCommitment> {
-        let root = self.tree.commit(&self.storage)?;
-        Ok(ClassCommitment(root))
+    pub fn commit_and_persist_changes(self) -> anyhow::Result<ClassCommitment> {
+        let update = self.tree.commit()?;
+        for (hash, node) in update.added {
+            self.storage.insert(&hash, &node)?;
```

Comment:
<purist>Function name `do_X_and_do_Y` just screams to be split in 2 separate functions.</purist>

If the tree still knows about concept of storage, this seems like loosing the coupling (while still keeping it), but not de-coupling completely. I was thinking that when de-coupled, the client code decides when to persist what.

I'd imagine keeping `MerkleTree` completely storage-unaware and adding something like `PersistentMerkelTree` wrapper (that has the same API but wraps tree and storage, and knows when to persist things) might be much easier to test and maintain.

Out of scope of this PR though.

Reply:
Agreed - but I didn't want to change all the call locations which previously relied on the tree doing this work implicitly.

`PersistentMerkelTree` - these wrappers are sort of like that already. But in a future PR I want the call sites to have more control over when they persist to storage. At which point this function will split into parts.